src/models/UndoRedo.ts

```diff
@@ -0,0 +1,75 @@
+import { INITIAL_SHOW_STATE, UNDO, REDO, UNDOABLE_ACTIONS } from "@/store/mutations"
+import { CalChartState } from "@/store";
+import Serializable from "./util/Serializable";
+
+// Increment upon making show metadata changes that break previous versions.
+const METADATA_VERSION = 1;
+
+// The UndoRedo holds the initial state of the show and all new changes
+// built on it.  It provides the long history of the show as it was created.
+// https://vuejsdevelopers.com/2017/11/13/vue-js-vuex-undo-redo/
+export class UndoRedo extends Serializable<UndoRedo> {
```

Comment:
Overall, I really like how you ended up doing this. I'm not sure how expensive it will be in practice to have an undo restore the initial state and then play back all of history minus the most recent action, but I like how unobtrusive this approach is. We don't need to change the way that we interact with the Vue store; the UndoRedo object just watches our interactions in the background to provide us with undo/redo functionality. I think this is pretty slick!

I think my biggest concern is how we will manage large histories. I imagine we'd want to bound the history at some point? As we move forward and forget the oldest elements in the history, I guess we will need to advance the INITIAL_SHOW_STATE? I'm interested to see how this shapes up; I know it's not quite done yet.

Reply:
I think that the 'unobtrusive' nature that I really like in this approach is also achievable using the alternative approach of 'taking a deep copy of the document state after each manipulation'. Given that this approach turned out to be pretty straightforward, I don't see any major benefits to taking the 'take a deep copy' approach. I think this has the potential to have a smaller footprint, and doesn't leave much room for difficult-to-track bugs (which was my initial fear before seeing these WIP changes). And because this new functionality is pretty well-confined to its own module/plugin, it would be pretty easy to replace with the 'deep copy' approach later if we ever discovered a compelling reason to use that method instead (and that gives me a sense of security in going forward with this approach).

Reply:
One thing that we could do is have something that reduces the length by making a new "initialState".  If we get to the undo length of say 1000, then each time something happens we create a new initial state and then remove the older states.